aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -15,8 +15,82 @@
 namespace at {
 namespace native {
 
+static inline Tensor integer_upcast(const Tensor& self, optional<ScalarType> dtype) {
+  ScalarType scalarType = self.type().scalarType();
+  ScalarType upcast_scalarType = dtype.value_or(at::isIntegralType(scalarType) ? ScalarType::Long : scalarType);
+  return self.toType(upcast_scalarType);
+}
+
+static inline Tensor cumsum(const Tensor& self, int64_t dim, optional<ScalarType> dtype) {
```

Comment:
It's a little funky that one of the overloads of cumsum is static inline, while the others aren't. Intentional?

Reply:
The static inline one is the one the other ones call, and what should be the real implementation once we hook up optional correctly.  The other ones are implementations of things defined in NativeFunctions.

Reply:
OK. It's confusing because the calls to this are fully qualified with a namespace (ostensibly to resolve an ambiguity) which is weird because normally if you want to call the local function, you don't qualify it.